daemon/emer-circular-file.c

```diff
@@ -0,0 +1,825 @@
+
+G_DEFINE_TYPE_WITH_CODE (EmerCircularFile, emer_circular_file, G_TYPE_OBJECT,
+                         G_ADD_PRIVATE (EmerCircularFile)
+                         G_IMPLEMENT_INTERFACE (G_TYPE_INITABLE, emer_circular_file_initable_init))
+
+enum
+{
+  PROP_0,
+  PROP_PATH,
+  PROP_MAX_SIZE,
+  NPROPS
+};
+
+static GParamSpec *emer_circular_file_props[NPROPS] = { NULL, };
+
+static gchar *
+get_metadata_filepath (EmerCircularFile *self)
+{
+  EmerCircularFilePrivate *priv =
+    emer_circular_file_get_instance_private (self);
+
+  gchar *data_filepath = g_file_get_path (priv->data_file);
+  gchar *metadata_filepath =
+    g_strconcat (data_filepath, METADATA_EXTENSION, NULL);
+  g_free (data_filepath);
+
+  return metadata_filepath;
+}
+
+static gboolean
+save_metadata_file (EmerCircularFile *self,
+                    GError          **error)
+{
+  EmerCircularFilePrivate *priv =
+    emer_circular_file_get_instance_private (self);
+
+  gchar *metadata_filepath = get_metadata_filepath (self);
+  gboolean save_succeeded =
+    g_key_file_save_to_file (priv->metadata_key_file, metadata_filepath, error);
+  g_free (metadata_filepath);
+
+  return save_succeeded;
+}
+
+static gboolean
+add_to_size (EmerCircularFile *self,
+             guint64           delta,
+             GError          **error)
+{
+  EmerCircularFilePrivate *priv =
+    emer_circular_file_get_instance_private (self);
+
+  guint64 new_size = priv->size + delta;
+  guint64 little_endian_size = swap_bytes_64_if_big_endian (new_size);
+  g_key_file_set_uint64 (priv->metadata_key_file, METADATA_GROUP_NAME, SIZE_KEY,
+                         little_endian_size);
+  gboolean save_succeeded = save_metadata_file (self, error);
+  if (save_succeeded)
+    priv->size = new_size;
+
+  return save_succeeded;
+}
+
+static gboolean
+set_metadata (EmerCircularFile *self,
+              guint64           size,
+              goffset           head,
+              GError          **error)
+{
+  EmerCircularFilePrivate *priv =
+    emer_circular_file_get_instance_private (self);
+
+  guint64 little_endian_size = swap_bytes_64_if_big_endian (size);
+  g_key_file_set_uint64 (priv->metadata_key_file, METADATA_GROUP_NAME, SIZE_KEY,
+                         little_endian_size);
+  guint64 little_endian_head = swap_bytes_64_if_big_endian (head);
+  g_key_file_set_int64 (priv->metadata_key_file, METADATA_GROUP_NAME, HEAD_KEY,
+                        little_endian_head);
+
+  if (!save_metadata_file (self, error))
+    return FALSE;
+
+  priv->size = size;
+  priv->head = head;
+  return TRUE;
+}
+
+static gboolean
+read_disk_bytes (EmerCircularFile *self,
+                 guint8           *buffer,
+                 gsize             num_bytes,
+                 gsize             max_size,
+                 GError          **error)
+{
+  EmerCircularFilePrivate *priv =
+    emer_circular_file_get_instance_private (self);
+
+  if (num_bytes == 0)
+    return TRUE;
```

Comment:
I don't think so. Presumably we aren't supposed to leave an arbitrarily large data file on disk if we resize to a smaller max size. It also seems counter-intuitive to me that if all of the following operations were successful, one would be left with an `EmerCircularFile` that still contained `bar`. We can't overwrite the data file without writing the metadata file either because that could result in the two being left in an inconsistent state.

```
EmerCircularFile *circular_file = emer_circular_file_new ("foo", 10, NULL);
emer_circular_file_append (circular_file, bar, 5);
g_object_unref (circular_file);
circular_file = emer_circular_file_new ("foo", 0, NULL); /* Write nothing because num_bytes == 0. */
g_object_unref (circular_file);
circular_file = emer_circular_file_new ("foo", 20, NULL);
```